File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefix(n, 0);
        vector<int> suffix(n, 0);
        vector<int> result(n,0);
        prefix[0] = nums[0];
        for(int i=1;i<n;i++){
            prefix[i] = prefix[i - 1] * nums[i];
        }
        suffix[n-1] = nums[n-1];
        for(int i=n-2;i>=0;i--){
            suffix[i] = suffix[i + 1] * nums[i];
        }
        for(int i=0;i<n;i++){
            if(i==0){
                result[i] = suffix[i+1];
            }
            else if(i==n-1){
                result[i] = prefix[n-2];
            }
            else{
                result[i] = prefix[i-1]*suffix[i+1];
            }
        }
        return result;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = [1] * n
        suffix = [1] * n
        res = [1] * n

        prefix[0] = nums[0]
        for i in range(1,n):
            prefix[i] = prefix[i-1] * nums[i]
        
        suffix[n-1] = nums[n-1]
        for i in range(n-2,-1,-1):
            suffix[i] = suffix[i+1] * nums[i]
        
        for i in range(n):
            if i == 0:
                res[i] = suffix[i+1]
            elif i == n-1:
                res[i] = prefix[i-1]
            else:
                res[i] = suffix[i+1] * prefix[i-1] 

        return res 